### PR TITLE
OSDOCS-2326: Updating understanding ingress topic

### PR DESCRIPTION
**Pull request by skrthomas**
https://issues.redhat.com/browse/OSDOCS-2326

Preview: https://deploy-preview-34657--osdocs.netlify.app/openshift-enterprise/latest/networking/understanding-networking.html

For 4.6+

Ready for QA: @lihongan @quarterpin @jechen0648

### CODE REVIEW

**Comment by lihongan** [comment]
cc @melvinjoseph86

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
+
+[id="nw-ne-comparing-ingress-route_{context}"]
+= Comparing Ingress and Routes
+The most common way to manage ingress traffic is with the Ingress Controller. OpenShift implements the Ingress Controller with a shared router service that runs as a pod inside the cluster. You can scale and replicate this pod like any other regular pod. This router service is based on the open source software HAProxy.
```
Comment (lihongan) [inline modules/nw-ne-comparing-ingress-route.adoc:7]: add link `http://www.haproxy.org/` to `open source software HAProxy` ?  WDYT?

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+Routes and ingress are the main resources for handling ingress traffic. An Ingress Controller is a Kubernetes resource that provides similar features to a route. Ingresses accept external requests and delegate them based on the route. You can only allow certain types of connections: HTTP/2, HTTPS and server name identification (SNI), and TLS with certificate. In OpenShift, routes are generated to meet the conditions specified by the ingress object.
```
Comment (lihongan) [inline modules/nw-ne-comparing-ingress-route.adoc:9]: should be `Routes and ingresses`

File: modules/nw-ne-ways-to-manage-ingress-traffic.adoc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+The following table provides an overview of ways that Administrators can manage their Ingress traffic:
+|===
+|Method |Why use?
+
+|Ingress
+|An ingress is a Kubernetes resource that provides some of the same features as routes (which are an OpenShift resource). Ingresses accept external requests and proxy them based on the route. You can only allow certain types of traffic: HTTP, HTTPS and server name identification (SNI), and TLS with SNI. In OpenShift, routes are generated to meet the conditions specified by the ingress object. Ingress managed LB
+Route
+
+|Route
+|Routes provide ingress traffic to services in the cluster. Routes were created before Kubernetes ingress objects and provide more features. Routes provide advanced features that may not be supported by Kubernetes ingress controllers through a standard interface, such as TLS re-encryption, TLS passthrough, and split traffic for blue-green deployments.
```
Comment (lihongan) [inline modules/nw-ne-ways-to-manage-ingress-traffic.adoc:16]: `Routes were created before Kubernetes ingress objects` is misleading statement, indeed the OCP created a concept called Route for the same purpose.

File: modules/nw-ne-comparing-ingress-route.adoc (hunk 1 above)
Comment (quarterpin) [reply modules/nw-ne-comparing-ingress-route.adoc:7]: Imo, The statement `This router service is based on the open source software HAProxy.` could be re-phrased as "This router service is based on HAproxy[insert link],  which is an opensource proxy/load balancer solution.

File: modules/nw-ne-comparing-ingress-route.adoc (hunk 2 above)
Comment (quarterpin) [reply modules/nw-ne-comparing-ingress-route.adoc:9]: I believe the statement `An Ingress Controller is a Kubernetes resource that provides similar features to a route` should be corrected to `Ingress is a Kubernetes resource that provides features similar to the route.  Ingress accepts external requests and delegates them based on the route. `

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+A route provides ingress traffic to services in the cluster. Routes were created before Kubernetes ingress objects and provide advanced features that may not be supported by standard Kubernetes ingress controllers, such as TLS re-encryption, TLS passthrough, and split traffic for blue-green deployments.
```
Comment (jechen0648) [inline modules/nw-ne-comparing-ingress-route.adoc:11]: Ingress traffic accesses service in the cluster through a route?

**Comment by lihongan** [comment]
cc @ShudiLi

File: modules/nw-ne-comparing-ingress-route.adoc (hunk 4 above)
Comment (skrthomas) [reply modules/nw-ne-comparing-ingress-route.adoc:11]: yes, great clarifying detail thank you! I agree it does fit at the beginning.

**Comment by skrthomas** [comment]
@lihongan @quarterpin @jeanchen Thanks a lot for your reviews :) I've incorporated your feedback.

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
+
+[id="nw-ne-comparing-ingress-route_{context}"]
+= Comparing Routes and ingresses
+The most common way to manage ingress traffic is with the Ingress Controller. OpenShift implements the Ingress Controller with a shared router service that runs as a pod inside the cluster. You can scale and replicate this pod like any other regular pod. This router service is based on link:http://www.haproxy.org/[HAProxy], which is an open source load balancer solution.
+
+Ingress traffic accesses services in the cluster through a route. Routes and ingresses are the main resources for handling ingress traffic. Ingress is a Kubernetes resource that provides features similar to a route. Ingress accepts external requests and delegates them based on the route. You can only allow certain types of connections: HTTP/2, HTTPS and server name identification (SNI), and TLS with certificate. In OpenShift, routes are generated to meet the conditions specified by the ingress object.
```
Comment (ctauchen) [inline modules/nw-ne-comparing-ingress-route.adoc:9]: Paragraphs 2 and 3 might be improved by making more explicit which resource you're talking about. The trouble is, both the ingress and route methods deal with ... ingress and routes, so it's a bit hard to tell the difference. 

Consider, for example, starting each paragraph with something along the lines of this:

1. The Kubernetes `Ingress` resource allows you to do this ...
2. The OpenShift `Route` resource allows you to do that ... 

That way, even if the text is a mix of ingress/route discussion, at least we've anchored a clear label on the reader's mind.

File: modules/nw-ne-ways-to-manage-ingress-traffic.adoc [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
+
+[id="nw-ne-ways-to-manage-ingress-traffic_{context}"]
+= Ways to manage ingress controller traffic
+The following table provides an overview of ways that Administrators can manage their Ingress traffic:
```
Comment (ctauchen) [inline modules/nw-ne-ways-to-manage-ingress-traffic.adoc:7]:
```suggestion
The following table provides an overview of ways that administrators can manage their ingress traffic:
```

**Review by ctauchen — COMMENTED** [review]
Nice to see improvements to the overview section! I've left a few comments for you to consider. Let me know if I can clarify anything.

**Comment by ctauchen** [comment]
@bergerhoffer @bobfuru I'd appreciate a second pass, if you've got the time. Thanks!

File: modules/nw-ne-comparing-ingress-route.adoc (hunk 5 above)
Comment (bergerhoffer) [inline modules/nw-ne-comparing-ingress-route.adoc:9]: * In OpenShift -> In {product-title}
* the ingress object -> the `Ingress` object ?

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+A route provides ingress traffic to services in the cluster. Routes were created before Kubernetes ingress objects and provide advanced features that may not be supported by standard Kubernetes ingress controllers, such as TLS re-encryption, TLS passthrough, and split traffic for blue-green deployments.
```
Comment (bergerhoffer) [inline modules/nw-ne-comparing-ingress-route.adoc:11]: * Same about being explicit when you're talking about `Route` objects and `Ingress` objects.
* may -> might

File: modules/nw-ne-openshift-ingress.adoc [hunk 8]
```diff
@@ -0,0 +1,10 @@
+// Module included in the following assemblies:
+// * understanding-networking.adoc
+
+
+[id="nw-ne-openshift-ingress_{context}"]
+= {product-title} Ingress Operator
+When you create your {product-title} cluster, pods and services running on the cluster are each allocated their own IP addresses. The IP addresses are accessible to other pods and services running nearby but not accessible to outside clients. The Ingress Operator implements the `ingresscontroller` API and is the component responsible for enabling external access to {product-title} cluster services.
```
Comment (bergerhoffer) [inline modules/nw-ne-openshift-ingress.adoc:7]: * but not -> but are not ?
* Should `ingresscontroller` API be capitalized differently, like `IngressController` API ?

File: modules/nw-ne-openshift-ingress.adoc [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+
+The Operator makes it possible for external clients to access your service by deploying and managing one or more HAProxy-based
```
Comment (bergerhoffer) [inline modules/nw-ne-openshift-ingress.adoc:9]: The Operator -> The Ingress Operator

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 10]
```diff
@@ -0,0 +1,11 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
+
+[id="nw-ne-comparing-ingress-route_{context}"]
+= Comparing Routes and ingresses
```
Comment (bergerhoffer) [inline modules/nw-ne-comparing-ingress-route.adoc:6]: In addition to what @ctauchen discusses, just wondering if you're trying to compare the actual objects or not. Because if so, then you could do Comparing Route and Ingress objects. (Headings are the only place we don't use backticks for object names.

And though some if this is definitely because I don't know much about ingress/routes, the explanation in this section isn't very clear to me what the differences are. I think it would be helpful to start out saying that one is a Kubernetes object and one is an openshift one. And then try to explain when you might want to use one over the other (if there are feature differences).

This is probably more detailed than you want or need here, but you could look at this for inspiration: https://deploy-preview-32226--osdocs.netlify.app/openshift-enterprise/latest/applications/deployments/what-deployments-are.html#deployments-comparing-deploymentconfigs_what-deployments-are

That compares Deployment and DeploymentConfig objects. But it right away tells you which is which, and what they share in common, and then where they differ.

File: modules/nw-ne-openshift-ingress.adoc [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+link:https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/[Ingress Controllers] to handle routing. You can use the Ingress Operator to route traffic by specifying {product-title} `Route` and Kubernetes `Ingress` resources. Configurations within the Ingress Controller, such as the ability to define `endpointPublishingStrategy` type and internal load balancing, provide ways to publish ingress controller endpoints.
```
Comment (bergerhoffer) [inline modules/nw-ne-openshift-ingress.adoc:10]: Should "ingress controller endpoints." be "Ingress Controller endpoints."?

File: modules/nw-ne-openshift-ways-to-manage-ingress-traffic.adoc [hunk 12]
```diff
@@ -0,0 +1,6 @@
+The following table provides an overview of ways that Administrators can manage their Ingress traffic:
+|===
+|Parameter |Description
+
+|`domain`
+|`domain`
```
[the remark this answers is not in the document]
Comment (bergerhoffer) [reply modules/nw-ne-openshift-ways-to-manage-ingress-traffic.adoc:6]: +1. And if not, it needs a title, etc. like a proper module and included somewhere

File: networking/understanding-networking.adoc [hunk 13]
```diff
@@ -4,13 +4,11 @@ include::modules/common-attributes.adoc[]
 :context: understanding-networking
 
 toc::[]
+As a cluster Administrator, you have several options for exposing applications that run inside a cluster to external traffic and securing network connections:
+* Service types, such as NodePort or load balancer
+* API resources, such as Ingress and route
```
[the remark this answers is not in the document]
Comment (bergerhoffer) [reply networking/understanding-networking.adoc:9]: Yeah, you could say `Ingress` and `Route` resources (or objects)

File: networking/understanding-networking.adoc [hunk 14]
```diff
@@ -4,13 +4,11 @@ include::modules/common-attributes.adoc[]
 :context: understanding-networking
 
 toc::[]
+As a cluster Administrator, you have several options for exposing applications that run inside a cluster to external traffic and securing network connections:
```
[the remark this answers is not in the document]
Comment (bergerhoffer) [reply networking/understanding-networking.adoc:7]: I know the "as a <user>" is the format for user stories, but I'm not personally a fan of actually writing this into the docs. Maybe consider saying "Cluster administrators have several options..."

File: modules/nw-ne-ways-to-manage-ingress-traffic.adoc [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+|===
+|Method |Why use?
```
Comment (bergerhoffer) [inline modules/nw-ne-ways-to-manage-ingress-traffic.adoc:9]: Suggest rewording "Why use?" Could say just "Use" or "Description" or something else.

File: modules/nw-ne-ways-to-manage-ingress-traffic.adoc [hunk 16]
```diff
@@ -0,0 +1,27 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
+
+[id="nw-ne-ways-to-manage-ingress-traffic_{context}"]
+= Ways to manage ingress controller traffic
+The following table provides an overview of ways that Administrators can manage their Ingress traffic:
+|===
```
Comment (bergerhoffer) [inline modules/nw-ne-ways-to-manage-ingress-traffic.adoc:8]: Suggest specifying column widths so that the second column isn't wrapping so much. Something like:

[cols="1,2",options="header"]

**Review by bergerhoffer — COMMENTED** [review]
Added a few additional thoughts!

File: modules/nw-ne-comparing-ingress-route.adoc [hunk 17]
```diff
@@ -0,0 +1,11 @@
+// Module included in the following assemblies:
+//
+// * ingress/understanding-networking.adoc
```
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/nw-ne-comparing-ingress-route.adoc:3]: I did mean networking/ apologies for this
